### PR TITLE
Making the node-to-node operator a nearest-point operator

### PR DESCRIPTION
**Pull request by sslattery**
The node-to-node operator can be made a nearest point operator if the check on coordinate equality is removed. If the meshes are the same, this still preserves the node-to-node transfer behavior. If they are different, the value at a given point is assigned to be the value of its nearest neighbor in other mesh.

### CODE REVIEW

**Comment by sslattery** [comment]
@aprokop Can you review?

**Comment by aprokop** [comment]
@sslattery ok.

**Comment by dalg24** [comment]
@sslattery Say I intend to do a node-to-node transfer but somehow the nodes are not perfectly aligned (e.g. translated a mesh wrong or used bad meshes) how would I then catch that with a nearest neighbor approach?

**Comment by sslattery** [comment]
We could add an option for node-to-node that activates the check for matching meshes.

**Comment by dalg24** [comment]
That would work.

**Comment by sslattery** [comment]
OK I added an option for the check and a unit test with a perturbation

File: packages/Operators/src/PointCloud/DTK_NodeToNodeOperator_impl.hpp [hunk 1]
```diff
@@ -137,12 +143,14 @@ void NodeToNodeOperator<DIM>::setupImpl(
...
+            // nodes are supposed to match.
+            DTK_REMEMBER( bool distance_check =
                 std::abs( EuclideanDistance<DIM>::distance(
-                    dist_sources( DIM * pairings.childCenterIds( i )[0], DIM )
-                        .getRawPtr(),
-                    target_centers( DIM * i, DIM ).getRawPtr() ) ) < 1.0e-14 );
+                              dist_sources( DIM * pairings.childCenterIds( i )[0], DIM )
+                              .getRawPtr(),
+                              target_centers( DIM * i, DIM ).getRawPtr() ) ) < 1.0e-14 );
+            DTK_CHECK( (distance_check && d_matching_nodes) || !d_matching_nodes );
```
Comment (sslattery) [inline packages/Operators/src/PointCloud/DTK_NodeToNodeOperator_impl.hpp]: I thought about that but didnt do it for two reasons:

1) I think separating it makes it more clear what check is being performed
2) This check is performed in DBC so I am not worried about performance.

**Comment by sslattery** [comment]
is this ok to merge?

**Comment by aprokop** [comment]
Looks good to me.

File: packages/Operators/test/tstNodeToNodeOperator.cpp [hunk 2]
```diff
@@ -173,12 +175,30 @@ void setupAndRunTest( const std::string &input_file,
...
+
+//---------------------------------------------------------------------------//
+TEUCHOS_UNIT_TEST( NodeToNodeOperator, non_matching_node_test )
 {
     // Run the test.
     Teuchos::Array<double> gold_data;
     Teuchos::Array<double> test_result;
-    setupAndRunTest( "node_to_node_test.xml", gold_data, test_result );
+    setupAndRunTest( "non_matching_node_test.xml", gold_data, test_result,
+                     true );
```
Comment (dalg24) [inline packages/Operators/test/tstNodeToNodeOperator.cpp]: Does `Teuchos` have something like `BOOST_CHECK_THROW` that would catch the `DataTransferKitException`?  We do not trigger that in the test.

Comment (dalg24) [inline packages/Operators/test/tstNodeToNodeOperator.cpp]: So I check and it does exists.  It is called [TEST_THROW](https://trilinos.org/docs/dev/packages/teuchos/doc/html/group__Teuchos__UnitTestAssertMacros__grp.html#ga8a6d1a22211e72a7eabe164cf3aa2103).  Please use it here and make sure the correct exception is thrown.

Comment (aprokop) [inline packages/Operators/test/tstNodeToNodeOperator.cpp]: The `TEST_THROW` should be hidden behind `#if HAVE_DTK_DBC`

Comment (sslattery) [inline packages/Operators/test/tstNodeToNodeOperator.cpp]: we could add something like that. technically to be very correct we should run all of our tests in scenarios that we expect the code to throw and check in a try/catch block

Comment (aprokop) [inline packages/Operators/test/tstNodeToNodeOperator.cpp]: Yes, there was a thought like that, but it would require a lot of effort.

Comment (dalg24) [inline packages/Operators/test/tstNodeToNodeOperator.cpp]: @sslattery No, look at the documentation for `TEST_THROW`.

Comment (sslattery) [inline packages/Operators/test/tstNodeToNodeOperator.cpp]: ill add a test for this quickly

File: packages/Operators/test/tstNodeToNodeOperator.cpp [hunk 3]
```diff
@@ -190,5 +211,28 @@ TEUCHOS_UNIT_TEST( NodeToNodeOperator, node_to_node_test )
...
+    Teuchos::Array<double> test_result;
+
+    bool caught_exception = false;
+    try
+    {
+        setupAndRunTest( "matching_node_test.xml", gold_data, test_result,
+                         true );
+    }
+    catch ( DataTransferKit::DataTransferKitException &e )
+    {
```
Comment (aprokop) [inline packages/Operators/test/tstNodeToNodeOperator.cpp]: I thought it should have been
```c++
#if HAVE_DTK_DBC
  TEST_THROW( setupAndRunTest( "matching_node_test.xml", gold_data, test_result, true ),
                                                            DataTransferKit::DataTransferKitException);
#endif
```
or something similar

**Comment by dalg24** [comment]
Where can we document `"Matching Nodes"`?

File: packages/Operators/test/tstNodeToNodeOperator.cpp [hunk 4]
```diff
@@ -190,5 +193,36 @@ TEUCHOS_UNIT_TEST( NodeToNodeOperator, node_to_node_test )
...
+}
+
+//---------------------------------------------------------------------------//
+TEUCHOS_UNIT_TEST( NodeToNodeOperator, exception_test )
+{
+    // Run the test assuming matching nodes but add a perturbation.
+    Teuchos::Array<double> gold_data;
+    Teuchos::Array<double> test_result;
+
+    bool caught_exception = false;
```
Comment (dalg24) [inline packages/Operators/test/tstNodeToNodeOperator.cpp]: This line is not needed any more.

File: packages/Operators/test/tstNodeToNodeOperator.cpp [hunk 5]
```diff
@@ -190,5 +193,36 @@ TEUCHOS_UNIT_TEST( NodeToNodeOperator, node_to_node_test )
...
+    TEST_THROW( setupAndRunTest( "matching_node_test.xml", gold_data,
+                                 test_result, true ),
+                DataTransferKit::DataTransferKitException );
+}
```
Comment (dalg24) [inline packages/Operators/test/tstNodeToNodeOperator.cpp]: I believe this need to be behind `#if HAVE_DTK_DBC`.

**Comment by dalg24** [comment]
@sslattery This was merged this into dtk-2.0.  Did we mean to pass it into master as well?

**Comment by sslattery** [comment]
i view this as a 2.0 update for CASL. The new code may want this functionality but we can add that when we refactor.

**Comment by aprokop** [comment]
I propose this to be merged in master. I think that master should not lack any features from dtk-2.0 as the original plan was for dtk-2.0 be like a fixed tag. Any changes that are suitable for dtk-2.0 should also go to the master, as otherwise when modifying master we'll have to compare against two baselines: earlier master **and** dtk-2.0, which is twice as much work.

**Pull request merged**